control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp

```diff
@@ -443,24 +450,59 @@ PidLongitudinalController::ControlData PidLongitudinalController::getControlData
     control_data.is_far_from_trajectory = true;
     return control_data;
   }
-  control_data.nearest_idx = nearest_idx;
+
+  // Delay compensation - Calculate the distance we got, predicted velocity and predicted
+  // acceleration after delay
+  control_data.state_after_delay =
+    predictedStateAfterDelay(control_data.current_motion, m_delay_compensation_time);
+
+  // calculate the target motion
+  constexpr double min_running_dist = 0.01;
+  if (control_data.state_after_delay.running_distance > min_running_dist) {
+    const auto target_pose = longitudinal_utils::findTrajectoryPoseAfterDistance(
+      control_data.nearest_idx, control_data.state_after_delay.running_distance,
+      control_data.interpolated_traj);
+    const auto target_interpolated_point =
+      calcInterpolatedTrajPointAndSegment(control_data.interpolated_traj, target_pose);
+    control_data.target_idx = target_interpolated_point.second + 1;
+    control_data.interpolated_traj.points.insert(
+      control_data.interpolated_traj.points.begin() + uint(control_data.target_idx),
+      target_interpolated_point.first);
+  }
+
+  // send debug values
+  m_debug_values.setValues(DebugValues::TYPE::PREDICTED_VEL, control_data.state_after_delay.vel);
+  m_debug_values.setValues(
+    DebugValues::TYPE::TARGET_VEL,
+    control_data.interpolated_traj.points.at(control_data.target_idx).longitudinal_velocity_mps);
 
   // shift
-  control_data.shift = getCurrentShift(control_data.nearest_idx);
+  control_data.shift = getCurrentShift(control_data);
   if (control_data.shift != m_prev_shift) {
     m_pid_vel.reset();
   }
   m_prev_shift = control_data.shift;
 
   // distance to stopline
   control_data.stop_dist = longitudinal_utils::calcStopDistance(
-    current_pose, m_trajectory, m_ego_nearest_dist_threshold, m_ego_nearest_yaw_threshold);
+    control_data.interpolated_traj.points.at(control_data.nearest_idx).pose,
+    control_data.interpolated_traj, m_ego_nearest_dist_threshold, m_ego_nearest_yaw_threshold);
 
   // pitch
   const double raw_pitch = longitudinal_utils::getPitchByPose(current_pose.orientation);
-  const double traj_pitch =
-    longitudinal_utils::getPitchByTraj(m_trajectory, control_data.nearest_idx, m_wheel_base);
-  control_data.slope_angle = m_use_traj_for_pitch ? traj_pitch : m_lpf_pitch->filter(raw_pitch);
+  const double traj_pitch = longitudinal_utils::getPitchByTraj(
+    control_data.interpolated_traj, control_data.target_idx, m_wheel_base);
+
+  // if vehicle is not in driving state, send the raw slope angle
```

Comment:
Regarding which slope information to use, it should not depend on the logic as a basic policy because it depends on the accuracy of sensors and maps. What is the purpose of this implementation?

Reply:
> Regarding which slope information to use, it should not depend on the logic as a basic policy because it depends on the accuracy of sensors and maps. What is the purpose of this implementation?



Actually, the pitch information from sensing would be more reliable from the trajectory. My purpose is, that if the vehicle is stopping, it can use the pitch information from sensing, with no need for future slope. But if the vehicle is moving (running distance after delay >0.0), it needs pitch information from the future.

While changing the pitch input source, to provide smoother change, I put them into LPF.

Maybe the slope input source should change with respect to running_distance threshold like:

if (running_distance > 0.5?) {
control_data.slope_angle = traj_pitch;
m_lpf_pitch->filter(control_data.slope_angle);
} else {
control_data.slope_angle = raw_pitch;
// send filtered slope angle to smooth transition between raw and traj pitch
control_data.slope_angle = m_lpf_pitch->filter(control_data.slope_angle);
}

Because in the current implementation, it is possible that the vehicle's velocity is < 0.001 but the vehicle may be in the DRIVE state. What do you think?